examples/sftp-server/main.go

```diff
@@ -0,0 +1,132 @@
+// An example SFTP server implementation using the golang SSH package.
+// Serves the whole filesystem visible to the user, and has a hard-coded username and password,
+// so not for real use!
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// Based on example server code from golang.org/x/crypto/ssh and server_standalone
+func main() {
+
+	readOnly := false
+	debugLevelStr := "none"
+	debugLevel := 0
+	debugStderr := false
+	rootDir := ""
+
+	flag.BoolVar(&readOnly, "R", false, "read-only server")
+	flag.BoolVar(&debugStderr, "e", false, "debug to stderr")
+	flag.StringVar(&debugLevelStr, "l", "none", "debug level")
+	flag.StringVar(&debugLevelStr, "root", "", "root directory")
+	flag.Parse()
+
+	// term := terminal.NewTerminal(channel, "> ")
+	debugStream := ioutil.Discard
+	if debugStderr {
+		debugStream = os.Stderr
+		debugLevel = 1
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	config := &ssh.ServerConfig{
+		PasswordCallback: func(c ssh.ConnMetadata, pass []byte) (*ssh.Permissions, error) {
+			// Should use constant-time compare (or better, salt+hash) in
+			// a production setting.
+			fmt.Fprintf(debugStream, "Login: %s\n", c.User())
+			if c.User() == "testuser" && string(pass) == "tiger" {
+				return nil, nil
+			}
+			return nil, fmt.Errorf("password rejected for %q", c.User())
+		},
+	}
+
+	privateBytes, err := ioutil.ReadFile("id_rsa")
+	if err != nil {
+		panic("Failed to load private key")
+	}
+
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		panic("Failed to parse private key")
+	}
+
+	config.AddHostKey(private)
+
+	// Once a ServerConfig has been configured, connections can be
+	// accepted.
+	listener, err := net.Listen("tcp", "0.0.0.0:2022")
+	if err != nil {
+		panic("failed to listen for connection")
+	}
+	fmt.Printf("Listening on %v\n", listener.Addr())
+
+	nConn, err := listener.Accept()
+	if err != nil {
+		panic("failed to accept incoming connection")
```

Comment:
Yes please! Old examples age like fruit in the sun. If you could please fix
up those as well that would be great.

On Sun, 8 Nov 2015, 10:22 Michael Houston <notifications@github.com> wrote:

> In examples/sftp-server/main.go
> <https://github.com/pkg/sftp/pull/54#discussion_r44218547>:
>
> > +		panic("Failed to parse private key")
> > +	}
> > +
> > +	config.AddHostKey(private)
> > +
> > +	// Once a ServerConfig has been configured, connections can be
> > +	// accepted.
> > +	listener, err := net.Listen("tcp", "0.0.0.0:2022")
> > +	if err != nil {
> > +		panic("failed to listen for connection")
> > +	}
> > +	fmt.Printf("Listening on %v\n", listener.Addr())
> > +
> > +	nConn, err := listener.Accept()
> > +	if err != nil {
> > +		panic("failed to accept incoming connection")
>
> This code is based on
> https://godoc.org/golang.org/x/crypto/ssh#example-NewServerConn - is it
> worth making these suggestions for inclusion there too?
>
> I agree that an example should demonstrate best-practice, especially if
> it's likely people will use it as the base for future work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pkg/sftp/pull/54/files#r44218547>.
>